<a href="https://colab.research.google.com/github/leearzuaga/Projects/blob/main/HackerNewsProject.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Exploring Hackers News Posts

##### In this project, I'll be analyzing two different types of posts from Hacker News, a popular website where technology related stories (or 'posts') are voted and commented upon (much like reddit). The two types of posts I'll be analyzing begin with either Ask HN or Show HN.

##### Ask HN are user-submitted posts that ask the Hacker News community a specific question like -- "What is the best online course you've ever taken?" Similarly, Show HN are user-submitted posts that show the Hacker News community a project, product, or generally just anything of interest.

##### I'll specifically compare these two types of posts to determine the following:

*   Do Ask HN or Show HN receive more comments on average?
*   Do posts created at a certain time receive more comments on average?

##### Please note that the data set I'm working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions.



### Introduction

In [ ]:
# Read in the hacker_news.csv file as a list of lists
import csv

f = open('hacker_news.csv')
hn = list(csv.reader(f))

# I'll display the first 5 rows to see what are list of lists data set looks like

hn[:5]



[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

### Removing Headers From Our Data Set

##### In order to analyze our data, we first need to remove the row containing the column headers from our data set.

In [ ]:
hn_headers = hn[0]

print(hn_headers)

hn = hn[1:]

print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


##### We can see from the above output that the data set contains the title of the posts, the number of comments for each post, and the date the post was created. Let's start by exploring the number of comments for each type of post

#### Extracting Ask HN and Show HN Posts

##### First, I'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [ ]:
# First I'll create 3 separate, empty lists -- one for each type of post

ask_posts = []
show_posts = [] 
other_posts = []

# Now we loop through our hn data set and assign the title of each post
# in each row to a variable named title

for post in hn:
  title = post[1]
  if title.lower().startswith('ask hn'):
    ask_posts.append(post)
  
  elif title.lower().startswith('show hn'):
    show_posts.append(post)

  else:
    other_posts.append(post)

# Checking out empty lists above to confirm our code was correct
print(ask_posts[:5])
print()
print(show_posts[:5])
print()
print(other_posts[:5])

# Finally we'll check the number of posts in ask_posts, show_posts, and other_posts
print()
print('Number Of Ask HN Posts : {num}'.format(num = len(ask_posts)))
print('Number Of Show HN Posts : {num}'.format(num = len(show_posts)))
print('Number Of Other HN Posts : {num}'.format(num = len(other_posts)))

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', '

##### We can see from the output above, each post was separated into their unique lists successfully. We can also can see from the above output the number of posts in each list varies with Other HN Posts being the most frequent type of posts.

### Calculating the Average Number of Comments for Ask HN and Show HN Posts

In [ ]:
# Finding the total number of comments in the ask posts list

total_ask_comments = 0

for post in ask_posts:
  total_ask_comments += int(post[4])

print('Total Ask Comments : {num}'.format(num = total_ask_comments))

# Finding the average number of ask comments
avg_ask_comments = total_ask_comments / len(ask_posts)

print('AVG Ask Comments : {num}'.format(num = avg_ask_comments))




Total Ask Comments : 24483
AVG Ask Comments : 14.038417431192661


In [ ]:
total_show_comments = 0

for post in show_posts:
  total_show_comments += int(post[4])

print('Total Show Comments : {num}'.format(num = total_show_comments))

avg_show_comments = total_show_comments / len(show_posts)

print('AVG Show Comments : {num}'.format(num = avg_show_comments))

Total Show Comments : 11988
AVG Show Comments : 10.31669535283993


##### We can see from the output that on average, Ask Posts recieve around 4 more comments per post than do Show Posts. Since ask posts are more likely to recieve comments, I'll focus my remaining analysis just on these posts.

### Finding the Amount of Ask Posts and Comments by Hour Created

##### Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [ ]:
# Calculating the amount of ask posts created during each hour of day and the number of comments received.

import datetime as dt

result_list = []

for post in ask_posts:
  created_at = post[6]
  num_comments = int(post[4])
  result_list.append([created_at, num_comments])


# print(result_list[:5])

counts_by_hour = {}
comments_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row in result_list:
  date = row[0]
  comment = row[1]
  time = dt.datetime.strptime(date, date_format).strftime('%H')

  if time not in counts_by_hour:
    counts_by_hour[time] = 1
    comments_by_hour[time] = comment

  else:
    counts_by_hour[time] += 1
    comments_by_hour[time] += comment

# print(counts_by_hour)
# print(comments_by_hour)
  



[['8/16/2016 9:55', 6], ['11/22/2015 13:43', 29], ['5/2/2016 10:14', 1], ['8/2/2016 14:20', 3], ['10/15/2015 16:38', 17]]
{'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}
{'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


##### After checking our output, we can see that we successfully retrieved the Ask Posts per hour and the number of comments per hour.

### Calculating the Average Number of Comments for Ask HN Posts by Hour

In [ ]:
# Calculating the average amount of comments Ask HN posts created at each hour of the day receive.
avg_by_hour = []

for hour in comments_by_hour:
  avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])

print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


### Sorting and Printing Values from a List of Lists

In [ ]:
swap_avg_by_hour = []

for row in avg_by_hour:
  swap_avg_by_hour.append([row[1], row[0]])

# print(swap_avg_by_hour)
sorted_swap = sorted(swap_avg_by_hour, reverse = True)

print("Top 5 Hours for 'Ask HN' Comments")
for avg, hour in sorted_swap[:5]:
  print('{} : {:.2f} average comments per post'.format(dt.datetime.strptime(hour, '%H').strftime('%H:%M'), avg))

Top 5 Hours for 'Ask HN' Comments
15:00 : 38.59 average comments per post
02:00 : 23.81 average comments per post
20:00 : 21.52 average comments per post
16:00 : 16.80 average comments per post
21:00 : 16.01 average comments per post


##### As we can see from our output, the hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments which to me is pretty significant.

### Conclusion

##### In this project, I analyzed Ask posts and Show posts on Hacker News website to determine which type of post and time receive the most comments on average. Based on my analysis, to maximize the amount of comments a post receives, I'd recommend the post be categorized as Ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

##### However, it should be noted that the data set I analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, Ask posts received more comments on average and Ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.